packages/ra-ui-materialui/src/button/DeleteWithConfirmButton.tsx

```diff
@@ -83,7 +83,7 @@ const DeleteWithConfirmButton: FC<DeleteWithConfirmButtonProps> = props => {
                             true
                         ),
                     }),
-                    id: record.id,
+                    id: record ? record.id : null,
```

Comment:
I agree with @Luwangel that the button should never be built without a record id, an that the responsibility to give the needed props props belong to his parent.
But maybe you can add a explic check in the button to alert explicitly when this id prop is missing.